File: .claude/worktrees/publish-fundsxml-book/examples/python/read_fund.py
#!/usr/bin/env python3
"""Chapter 12, Section 12.3.2 — Read a FundsXML file and print the fund LEI and total NAV.
Run: python3 read_fund.py ../europa_growth_fund.xml
Requires: pip install lxml
"""
import sys
from lxml import etree

def read_fund(path: str) -> None:
    doc = etree.parse(path)
    fund = doc.find("./Funds/Fund")
    if fund is None:
        print(f"{path}: no Fund element found", file=sys.stderr)
        sys.exit(1)

    lei = fund.findtext("Identifiers/LEI") or "(no LEI)"
    name = fund.findtext("Names/OfficialName") or "(unnamed)"

    tav = fund.find(
        "FundDynamicData/TotalAssetValues/TotalAssetValue/"
        "TotalNetAssetValue/Amount"
    )
    if tav is None:
        print(f"{name} ({lei}): no TotalNetAssetValue found")
        return

    amount = tav.text
    currency = tav.get("ccy")
    print(f"{name}")
    print(f"  LEI:  {lei}")
    print(f"  NAV:  {amount} {currency}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: read_fund.py <file.xml>", file=sys.stderr)
        sys.exit(2)
    read_fund(sys.argv[1])

File: .claude/worktrees/publish-fundsxml-book/examples/java/ReadFund.java
// Chapter 12, Section 12.3.3 — Read a FundsXML file with JAXP + XPath and print LEI + NAV.
// Compile and run: javac ReadFund.java && java ReadFund ../europa_growth_fund.xml
import java.io.File;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Node;

public class ReadFund {
    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            System.err.println("usage: java ReadFund <file.xml>");
            System.exit(2);
        }

        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        dbf.setNamespaceAware(false);
        DocumentBuilder db = dbf.newDocumentBuilder();
        Document doc = db.parse(new File(args[0]));

        XPath xpath = XPathFactory.newInstance().newXPath();
        Node fund = (Node) xpath.evaluate("/FundsXML4/Funds/Fund", doc,
                javax.xml.xpath.XPathConstants.NODE);
        if (fund == null) {
            System.err.println(args[0] + ": no Fund element found");
            System.exit(1);
        }

        String lei = xpath.evaluate("Identifiers/LEI", fund);
        String name = xpath.evaluate("Names/OfficialName", fund);
        Node amt = (Node) xpath.evaluate(
            "FundDynamicData/TotalAssetValues/TotalAssetValue/"
            + "TotalNetAssetValue/Amount", fund,
            javax.xml.xpath.XPathConstants.NODE);

        if (amt == null) {
            System.out.println(name + " (" + lei + "): no TotalNetAssetValue");
            return;
        }

        String amount = amt.getTextContent();
        String currency = amt.getAttributes().getNamedItem("ccy").getNodeValue();

        System.out.println(name);
        System.out.println("  LEI:  " + lei);
        System.out.println("  NAV:  " + amount + " " + currency);
    }
}
